File: agentos/auth/clerk.py
"""Clerk token verification helpers (RS256 via JWKS).

This module lets AgentOS trust Clerk-issued JWTs without introducing
additional JWT libraries. It verifies signature, expiry, issuer, and
optionally audience against environment configuration.
"""

from __future__ import annotations

import base64
import json
import os
import time
from dataclasses import dataclass
from typing import Any

import httpx


@dataclass
class ClerkClaims:
    sub: str
    email: str
    name: str
    org_id: str
    org_name: str
    org_role: str
    iss: str
    exp: int
    iat: int
    aud: str | None = None


def clerk_enabled() -> bool:
    return os.environ.get("AGENTOS_AUTH_PROVIDER", "").lower() == "clerk"


def _b64url_decode(value: str) -> bytes:
    padded = value + "=" * ((4 - len(value) % 4) % 4)
    return base64.urlsafe_b64decode(padded)


def _issuer() -> str:
    issuer = os.environ.get("AGENTOS_CLERK_ISSUER", "").strip()
    return issuer.rstrip("/")


def _audience() -> str:
    return os.environ.get("AGENTOS_CLERK_AUDIENCE", "").strip()


def _jwks_url() -> str:
    configured = os.environ.get("AGENTOS_CLERK_JWKS_URL", "").strip()
    if configured:
        return configured
    issuer = _issuer()
    if not issuer:
        raise ValueError("AGENTOS_CLERK_ISSUER is required for Clerk auth")
    return f"{issuer}/.well-known/jwks.json"


def _load_jwks() -> dict[str, Any]:
    url = _jwks_url()
    timeout = float(os.environ.get("AGENTOS_CLERK_JWKS_TIMEOUT_SECONDS", "5"))
    response = httpx.get(url, timeout=timeout)
    response.raise_for_status()
    payload = response.json()
    if not isinstance(payload, dict) or "keys" not in payload:
        raise ValueError("Invalid JWKS payload from Clerk")
    return payload


def _public_key_from_jwk(jwk: dict[str, Any]):
    from cryptography.hazmat.primitives.asymmetric import rsa

    n = int.from_bytes(_b64url_decode(jwk["n"]), "big")
    e = int.from_bytes(_b64url_decode(jwk["e"]), "big")
    return rsa.RSAPublicNumbers(e, n).public_key()


def _pick_key(jwks: dict[str, Any], kid: str) -> dict[str, Any] | None:
    keys = jwks.get("keys", [])
    for key in keys:
        if key.get("kid") == kid and key.get("kty") == "RSA":
            return key
    return None


def verify_clerk_token(token: str) -> ClerkClaims | None:
    """Verify Clerk JWT and return normalized claims."""
    try:
        from cryptography.hazmat.primitives import hashes
        from cryptography.hazmat.primitives.asymmetric import padding

        header_b64, payload_b64, signature_b64 = token.split(".")
        header = json.loads(_b64url_decode(header_b64))
        payload = json.loads(_b64url_decode(payload_b64))
        if header.get("alg") != "RS256":
            return None
        kid = header.get("kid", "")
        if not kid:
            return None

        jwks = _load_jwks()
        jwk = _pick_key(jwks, kid)
        if jwk is None:
            return None

        pub_key = _public_key_from_jwk(jwk)
        signed = f"{header_b64}.{payload_b64}".encode()
        signature = _b64url_decode(signature_b64)
        pub_key.verify(signature, signed, padding.PKCS1v15(), hashes.SHA256())

        now = int(time.time())
        exp = int(payload.get("exp", 0))
        iat = int(payload.get("iat", 0))
        if exp <= now or iat > now + 60:
            return None

        expected_iss = _issuer()
        if expected_iss and payload.get("iss", "").rstrip("/") != expected_iss:
            return None

        expected_aud = _audience()
        if expected_aud:
            aud = payload.get("aud")
            if isinstance(aud, list):
                if expected_aud not in aud:
                    return None
            elif aud != expected_aud:
                return None

        email = payload.get("email", "") or payload.get("primary_email_address", "")
        name = payload.get("name", "") or payload.get("full_name", "") or email.split("@")[0]
        org_id = (
            payload.get("org_id")
            or payload.get("organization_id")
            or payload.get("org")
            or ""
        )
        org_name = payload.get("org_name") or payload.get("organization_name") or ""
        org_role = (
            payload.get("org_role")
            or payload.get("organization_role")
            or payload.get("role")
            or ""
        )
        return ClerkClaims(
            sub=str(payload.get("sub", "")),
            email=str(email),
            name=str(name),
            org_id=str(org_id),
            org_name=str(org_name),
            org_role=str(org_role),
            iss=str(payload.get("iss", "")),
            exp=exp,
            iat=iat,
            aud=str(payload.get("aud")) if payload.get("aud") else None,
        )
    except Exception:
        return None

File: control-plane/src/auth/clerk.ts
/**
 * Clerk RS256 JWKS verification — ported from agentos/auth/clerk.py.
 * Fetches JWKS from Clerk's well-known endpoint, verifies RS256 signatures.
 */
import type { TokenClaims } from "./types";

interface JWKSCache {
  keys: JsonWebKey[];
  fetchedAt: number;
}

let jwksCache: JWKSCache | null = null;
const JWKS_TTL_MS = 5 * 60 * 1000; // 5 minutes

async function fetchJwks(issuer: string, jwksUrl?: string): Promise<JsonWebKey[]> {
  if (jwksCache && Date.now() - jwksCache.fetchedAt < JWKS_TTL_MS) {
    return jwksCache.keys;
  }

  const url = jwksUrl || `${issuer.replace(/\/$/, "")}/.well-known/jwks.json`;
  const resp = await fetch(url, { headers: { Accept: "application/json" } });
  if (!resp.ok) throw new Error(`JWKS fetch failed: ${resp.status}`);

  const data = (await resp.json()) as { keys: JsonWebKey[] };
  jwksCache = { keys: data.keys, fetchedAt: Date.now() };
  return data.keys;
}

async function importRsaKey(jwk: JsonWebKey): Promise<CryptoKey> {
  return crypto.subtle.importKey(
    "jwk",
    jwk,
    { name: "RSASSA-PKCS1-v1_5", hash: "SHA-256" },
    false,
    ["verify"],
  );
}

function b64urlDecode(s: string): Uint8Array {
  let padded = s.replace(/-/g, "+").replace(/_/g, "/");
  const pad = 4 - (padded.length % 4);
  if (pad !== 4) padded += "=".repeat(pad);
  const binary = atob(padded);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

export function clerkEnabled(issuer?: string): boolean {
  return !!issuer;
}

export function deriveDisplayName(payload: Record<string, unknown>): string {
  const explicitName = String(payload.name ?? "").trim();
  if (explicitName) return explicitName;

  const first = String(payload.given_name ?? "").trim();
  const last = String(payload.family_name ?? "").trim();
  const joined = [first, last].filter(Boolean).join(" ").trim();
  if (joined) return joined;

  const email = String(payload.email ?? payload.primary_email_address ?? "").trim().toLowerCase();
  if (email && email.includes("@")) {
    const localPart = email.split("@")[0].replace(/[._-]+/g, " ").trim();
    if (localPart) return localPart;
  }
  return "";
}

export async function verifyClerkToken(
  token: string,
  issuer: string,
  opts?: { audience?: string; jwksUrl?: string },
): Promise<TokenClaims | null> {
  try {
    const parts = token.split(".");
    if (parts.length !== 3) return null;

    const [headerB64, payloadB64, sigB64] = parts;

    // Parse header for kid
    const header = JSON.parse(new TextDecoder().decode(b64urlDecode(headerB64)));
    if (header.alg !== "RS256") return null;

    // Find matching JWK
    const keys = await fetchJwks(issuer, opts?.jwksUrl);
    const jwk = header.kid ? keys.find((k) => (k as any).kid === header.kid) : keys[0];
    if (!jwk) return null;

    // Verify signature
    const key = await importRsaKey(jwk);
    const signingInput = new TextEncoder().encode(`${headerB64}.${payloadB64}`);
    const signature = b64urlDecode(sigB64);
    const valid = await crypto.subtle.verify("RSASSA-PKCS1-v1_5", key, signature, signingInput);
    if (!valid) return null;

    // Parse payload
    const payload = JSON.parse(new TextDecoder().decode(b64urlDecode(payloadB64)));

    // Verify issuer (strip trailing slashes from both sides)
    if (payload.iss) {
      const normalizedIss = String(payload.iss).replace(/\/+$/, "");
      const normalizedIssuer = issuer.replace(/\/+$/, "");
      if (normalizedIss !== normalizedIssuer) return null;
    }

    // Verify audience (handle both string and array)
    if (opts?.audience) {
      const aud = payload.aud;
      if (Array.isArray(aud)) {
        if (!aud.includes(opts.audience)) return null;
      } else if (aud !== opts.audience) {
        return null;
      }
    }

    // Check expiry
    if (payload.exp && payload.exp < Math.floor(Date.now() / 1000)) return null;

    // Reject tokens with iat too far in the future (clock skew tolerance: 60s)
    const nowSec = Math.floor(Date.now() / 1000);
    if (payload.iat && payload.iat > nowSec + 60) return null;

    return {
      sub: payload.sub ?? "",
      email: payload.email ?? payload.primary_email_address ?? "",
      name: deriveDisplayName(payload),
      provider: "clerk",
      org_id: payload.org_id ?? payload.organization_id ?? "",
      iat: payload.iat ?? 0,
      exp: payload.exp ?? 0,
      role: payload.org_role ?? payload.organization_role ?? "",
    };
  } catch {
    return null;
  }
}
